src/protocols/cosmic_workspaces/cosmic-workspaces.c

```diff
@@ -212,17 +213,27 @@ workspace_set_state(struct lab_cosmic_workspace *workspace,
 }
 
 /* Group */
+static void
+group_handle_transaction_destroy(struct wl_listener *listener, void *data)
+{
+	struct lab_transaction *transaction = data;
+	wlr_log(WLR_ERROR, "destroying group transaction data %p", transaction->data);
+	free(transaction->data);
+}
+
 static void
 group_handle_create_workspace(struct wl_client *client,
 		struct wl_resource *resource, const char *name)
 {
-	struct wl_resource_addon *addon = wl_resource_get_user_data(resource);
+	struct lab_wl_resource_addon *addon = wl_resource_get_user_data(resource);
 	if (!addon) {
 		return;
 	}
 
 	struct lab_cosmic_workspace_group *group = addon->data;
-	transaction_add_workspace_group_ev(group, resource, CW_PENDING_WS_CREATE, name);
+	struct lab_transaction *transaction = lab_transaction_add(
+		addon->ctx, CW_PENDING_WS_CREATE, group, xstrdup(name));
+	wl_signal_add(&transaction->events.destroy, &group->on.transaction_destroy);
```

Comment:
Hrm. This will leak when there are multiple create-workspace transactions within a single `commit()`. Maybe revert the whole destroy handler for the transaction idea and instead go back to manually `free(->data)` in `manager_handle_commit()`. Same issue for cosmic-workspaces.